plugins/modules/ec2_vpc_vgw.py

```diff
@@ -325,95 +344,92 @@ def find_vgw(client, module, vpn_gateway_id=None):
-    # found and we will not create another vgw.
+def ensure_vgw_present(client, module: AnsibleAWSModule) -> Tuple[bool, Dict[str, Any]]:
+    changed = False
+    params = {
+        "Name": module.params.get("name"),
+        "VpcId": module.params.get("vpc_id"),
+        "VpnGatewayIds": module.params.get("vpn_gateway_id"),
+        "Tags": module.params.get("tags") or {},
+        "PurgeTags": module.params.get("purge_tags", False),
+    }
+
+    existing_vgw = (
+        find_vgw(client, module, [params["VpnGatewayIds"]]) if params["VpnGatewayIds"] else find_vgw(client, module)
+    )
+
+    if existing_vgw:
+        changed |= handle_existing_vgw(client, module, existing_vgw[0], params)
+    else:
+        changed |= create_new_vgw(client, module, params)
+
+    vgw = find_vgw(client, module, [existing_vgw[0]["VpnGatewayId"]]) if existing_vgw else find_vgw(client, module)
+    return changed, get_vgw_info(vgw)
 
+
+def handle_existing_vgw(client, module: AnsibleAWSModule, existing_vgw: dict, params: dict) -> bool:
     changed = False
-    params = dict()
-    result = dict()
-    params["Name"] = module.params.get("name")
-    params["VpcId"] = module.params.get("vpc_id")
-    params["Type"] = module.params.get("type")
-    params["Tags"] = module.params.get("tags")
-    params["VpnGatewayIds"] = module.params.get("vpn_gateway_id")
+    vpn_gateway_id = existing_vgw["VpnGatewayId"]
+
+    # Update tags
+    desired_tags = module.params.get("tags")
+    purge_tags = module.params.get("purge_tags")
+    if desired_tags is None:
+        desired_tags = dict()
+        purge_tags = False
+    tags = dict(Name=module.params.get("name"))
+    tags.update(desired_tags)
+    changed |= ensure_ec2_tags(
+        client, module, vpn_gateway_id, resource_type="vpn-gateway", tags=tags, purge_tags=purge_tags
+    )
 
-    # check that the vpc_id exists. If not, an exception is thrown
+    # Manage VPC attachments
+    current_vpc_attachments = existing_vgw["VpcAttachments"]
     if params["VpcId"]:
-        vpc = find_vpc(client, module)
+        if current_vpc_attachments and current_vpc_attachments[0]["State"] == "attached":
+            current_vpc_id = current_vpc_attachments[0]["VpcId"]
+            if current_vpc_id != params["VpcId"]:
+                if module.check_mode:
+                    return True
+                detach_vgw(client, module, vpn_gateway_id, current_vpc_id)
+                get_waiter(client, "vpn_gateway_detached").wait(VpnGatewayIds=[vpn_gateway_id])
+                attach_vgw(client, module, vpn_gateway_id)
+                changed = True
+        else:
+            attach_vgw(client, module, vpn_gateway_id)
+            changed = True
+    else:
+        if current_vpc_attachments and current_vpc_attachments[0]["State"] == "attached":
+            if module.check_mode:
+                return True
+            detach_vgw(client, module, vpn_gateway_id, current_vpc_attachments[0]["VpcId"])
+            changed = True
```

Comment:
```suggestion
```
The VPC should not be detached the Gateway, here nothing has to be done has the VPC was not specified and state is set to present. The VPC is detached when state is absent and VPC id is provided

Reply:
I'm not sure if that's correct. The current integration tests indicate that if
`state = present` and `vpc_id` is `not specified` then the `vgw` will be `detached` from the `vpcs`.

These tests are pre-existing, so module behavior is that before I started with refactoring
```
    - name: Detach vpn gateway
      community.aws.ec2_vpc_vgw:
        state: present
        name: "{{ vgw_name }}"
      register: vgw

    - name: Assert results
      ansible.builtin.assert:
        that:
          - vgw.changed
          - not vgw.vgw.vpc_id
   ```

similarly when `state = absent` then `vgw` will be `deleted` (by first detaching it from vpc)
```
    - name: Delete vpn gateway
      community.aws.ec2_vpc_vgw:
        state: absent
        vpn_gateway_id: '{{ vgw.vgw.id | default(vgw_id) }}'
      ignore_errors: true
```